src/TreeSitter/Importing.hs

```diff
@@ -1,135 +1,271 @@
-      deriving (Eq, Ord, Show)
+-- Parse source code and produce AST
+parseByteString :: Building t => Ptr TS.Language -> ByteString -> IO (Either String t)
+parseByteString language bytestring = withParser language $ \ parser -> withParseTree parser bytestring $ \ treePtr ->
+  if treePtr == nullPtr then
+    pure (Left "error: didn't get a root node")
+  else
+    withRootNode treePtr $ \ rootPtr ->
+      withCursor (castPtr rootPtr) $ \ cursor ->
+        runM (runFail (runReader cursor (runReader bytestring buildNode)))
 
-data Number = Number
-      deriving (Eq, Ord, Show)
+-- | Building is the process of iterating over tree-sitter’s parse trees using its tree cursor API and producing Haskell ASTs for the relevant nodes.
+--
+--   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'buildNode' providing that they have a suitable 'Generic' instance.
+class Building a where
+  buildNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  default buildNode :: (MonadFail m, Carrier sig m, GBuilding (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  buildNode = to <$> gbuildNode
 
-data Identifier = Identifier
-      deriving (Eq, Ord, Show)
+  buildEmpty :: MonadFail m => m a
+  buildEmpty = fail "expected a node"
 
+instance Building Text.Text where
+  buildNode = do
+    node <- peekNode
+    case node of
+      Just node' -> do
+        bytestring <- ask
+        let start = fromIntegral (nodeStartByte node')
+            end = fromIntegral (nodeEndByte node')
+        pure (decodeUtf8 (slice start end bytestring))
+      _ -> fail "expected a node for Text"
 
-importByteString :: (Importing t) => Ptr TS.Parser -> ByteString -> IO (Maybe t)
-importByteString parser bytestring =
-  unsafeUseAsCStringLen bytestring $ \ (source, len) -> alloca (\ rootPtr -> do
-      let acquire =
-            ts_parser_parse_string parser nullPtr source len
+instance Building a => Building (Maybe a) where
+  buildNode = Just <$> buildNode
+  buildEmpty = pure Nothing
 
-      let release t
-            | t == nullPtr = pure ()
-            | otherwise = ts_tree_delete t
+instance (Building a, Building b, SymbolMatching a, SymbolMatching b) => Building (Either a b) where
+  buildNode = do
+      currentNode <- peekNode
+      (lhsSymbolMatch, rhsSymbolMatch, currentNode) <- case currentNode of
+        Just node -> pure (symbolMatch (Proxy @a) node, symbolMatch (Proxy @b) node, node)
+        Nothing -> fail "expected a node; didn't get one"
+      if lhsSymbolMatch -- FIXME: report error
+        then Left <$> buildNode @a
+        else if rhsSymbolMatch
+          then Right <$> buildNode @b
+          else fail $ showFailure (Proxy @a) currentNode `sep` showFailure (Proxy @b) currentNode -- TODO: do the toEnum nodeSymbol stuff for the current node to show the symbol name
```

Comment:
1. This `TODO` is obsolete.

2. We could simplify this by using the `SymbolMatching` instance for `Either` instead.

```suggestion
          else fail $ showFailure (Proxy @(Either a b)) currentNode
```